R/common-impute.R

```diff
@@ -48,33 +48,95 @@ groupedLOCF <- function(df, id) {
 #' @seealso \code{\link{healthcareai}}
 #' @examples
-#' # For a numeric vector
-#' vResult <- imputeColumn(c(1,2,3,NA))
-#'
-#' # For a factor vector
-#' vResult <- imputeColumn(c('Y','N','Y',NA))
-#'
-#' # To use this function on an entire data frame:
-#' df <- data.frame(a=c(1,2,3,NA),
-#'                 b=c('Y','N','Y',NA))
-#' df[] <- lapply(df, imputeColumn)
-#' head(df)
-imputeColumn <- function(v) {
-  if (is.numeric(v)) {
-    v[is.na(v)] <- mean(v, na.rm = TRUE)
+#' 
+#' # Impute a single column
+#' df <- data.frame(a=c(1,2,3,NA), b=c('Y','N','Y',NA),
+#'    c=c(11,21,31,43), d=c('Y','N','N',NA))
+#' df <- df['a'] # note df[,1] does not return a df!
+#' out <- imputeDF(df)
+#' dfOut <- out$df # imputed data frame
+#' imputeVals <- out$imputeVals # imputed values
+#' print(dfOut)
+
+#' # Impute an entire data frame
+#' df <- data.frame(a=c(1,2,3,NA), b=c('Y','N','Y',NA),
+#'    c=c(11,21,31,43), d=c('Y','N','N',NA))
+#' out <- imputeDF(df)
+#' dfOut <- out$df # imputed data frame
+#' imputeVals <- out$imputeVals # imputed values
+#' print(dfOut)
+#' 
+#' # To impute using your own values (one per column)
+#' df <- data.frame(a=c(1,2,3,NA), b=c('Y','N','Y',NA),
+#'    c=c(11,21,31,43), d=c('Y','N','N',NA))
+#' myValues <- list(2, 'Y', 26.5, 'N') 
+#' out <- imputeDF(df, myValues)
+#' dfOut <- out$df # imputed data frame
+#' print(dfOut)
+
+imputeDF <- function(df, imputeVals = list()) {
+  # Use separate functions to make applying easier in deploy
+  # This function finds the value to impute for a column
+  getImputeValuesColumn <- function(col) {
+    if (is.numeric(col)) {
+      value <- mean(col, na.rm = TRUE) # this is where the method could change
+    } else {
+      value <- names(which.max(table(col)))
+    }
+    return(value)
+  }
+  
+  # This function applies the value to a column
+  applyImputeValuesColumn <- function(col, val) {
+    col[is.na(col)] <- val
+    return(col)
+  }
+
+  # Check to make sure that df is a dataframe
+  if (!(is.data.frame(df))) {
+    stop('df must be a dataframe.')
+  }
+
+  # Change factor columns to characters so that missing levels are dealt with correctly.
+  for (col in names(df)) {
+    if (is.factor(df[[col]])) {
+      df[[col]] <- as.character(df[[col]])
+    }
+  }
+
+  # Calculate imputation values if needed
+  if (length(imputeVals) == 0) {
+    # Calculate values
+    imputeVals <- lapply(df, getImputeValuesColumn)
   } else {
-    v[is.na(v)] <- names(which.max(table(v)))
+    if (!(is.list(imputeVals))){
+      stop("imputeValues must be a list.")
+    }
+    # cat('Using supplied values for imputation.', '\n')
+    if (length(imputeVals) != ncol(df)) {
+      stop('Your dataframe must have the same number of columns as your provided list!')
+    }
   }
-  v
-}
\ No newline at end of file
+
+  # Apply imputation values
+  df[] <- lapply(1:ncol(df), function(x,y) applyImputeValuesColumn(df[,x], imputeVals[[x]]))
+  
+  # Change characters back to factors.
```

Comment:
Do we need to worry about the case where the dataframe passed to imputeDF has character columns instead of factor columns (if someone is calling imputeDF outside of supervised model dev/deploy)?